truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedCallTarget.java

```diff
@@ -855,6 +858,60 @@ final boolean isCompilationFailed() {
         return compilationFailed;
     }
 
+    final boolean isCompilationStopped() {
+        OptimizedTruffleRuntime.CompilationActivityMode compilationActivityMode = runtime().getCompilationActivityMode();
+        long sct = runtime().stoppedCompilationTime().get();
+        if (compilationActivityMode == OptimizedTruffleRuntime.CompilationActivityMode.STOP_COMPILATION) {
+            if (sct != 0 && System.currentTimeMillis() - sct > engine.stoppedCompilationRetryDelay) {
+                runtime().stoppedCompilationTime().compareAndSet(sct, 0);
+                // Try again every StoppedCompilationRetryDelay milliseconds to potentially trigger a code cache sweep.
+                compilationActivityMode = OptimizedTruffleRuntime.CompilationActivityMode.RUN_COMPILATION;
+            }
+        }
+
+        switch (compilationActivityMode) {
+            case RUN_COMPILATION : {
+                // This is the common case - compilations are not stopped.
+                return false;
+            }
+            case STOP_COMPILATION : {
+                if (sct == 0) {
+                    runtime().stoppedCompilationTime().compareAndSet(0, System.currentTimeMillis());
+                }
+                // Flush the compilations queue for now. There's still a chance that compilation
+                // will be re-enabled eventually, if the hosts code cache can be cleaned up.
+                Collection<OptimizedCallTarget> targets = runtime().getCompileQueue().getQueuedTargets(null);
+                // If there's just a single compilation target in the queue, the chance is high that it is
+                // the one we've just added after the StoppedCompilationRetryDelay ran out, so keep it to
+                // potentially trigger a code cache sweep.
+                if (targets.size() > 1) {
```

Comment:
> the one we've just added 

I assume if this assumption is false, the compilation would fail since it has been stopped, right?
You basically cancel the compilations to avoid having them all fail?
What kind of failure is to be expected if it is stopped? A non permanent bailout? Any other failure could trigger OptimizedCallTarget.compilationFailed to be set to true, which would no longer call this method.

Reply:
This code is based on the characteristics of the current HotSpot code cache sweeping. After [JDK-8290025: Remove the Sweeper](https://bugs.openjdk.org/browse/JDK-8290025) (i.e. since JDK 20), code cache sweeping can only be triggered by a full GC if certain conditions like the code cache allocation rate or the code cache fill state meet certain criteria. Each time before a new nmethod will be inserted into the code cache, these criteria are checked (see [`CodeCache::gc_on_allocation()`](https://github.com/openjdk/jdk/blob/ed03f0d9d10518242a3dc6e3685f1bdb0550c723/src/hotspot/share/code/codeCache.cpp#L776)) and if the criteria is met, a full GC will be triggered which may in turn trigger a sweep of the code cache. The nmethod age (and thus the condition on when nmethods will be removed from the code cache) is determined by the number of last such sweeping cycles they have been active.

The current algorithm is not perfect. E.g. once compilation gets disabled, there will be no more code cache checks due to `CodeCache::gc_on_allocation()`. Also, the code cache allocation rate will be zero after compilation has been disabled. Finally, if the application is not running full GCs because of a full heap for a long time (which is not uncommon for a well-tuned application), no more code cache sweeping will be triggered at all.

I plan to address some of these shortcoming, but for now it looked beneficial in my experiments, to periodically trigger a Truffle compilation (and thus a call to `CodeCache::gc_on_allocation()`), even if compilation is disabled. That's why I introduced the `StoppedCompilationRetryDelay` mechanism in the first place.

But in my experiments I observed that once I enter a new compilation into the queue because the `StoppedCompilationRetryDelay`  passed, in most cases it was almost immediately removed from the queue again before it could be picked up by a compiler thread because the next invocation of `OptimizedCallTarget::compile()` was just faster (i.e. it took the `STOP_COMPILATION` path and drained the queue). This means that the `StoppedCompilationRetryDelay` mechanism, in contrast to its design, almost never triggered a compilation.

The assumption in the comment, that a single entry in the queue was probably entered because `StoppedCompilationRetryDelay` timed out is only wrong at the beginning, when compilation gets stopped (i.e. the compilation activity mode changed from `RUN_COMPILATION` to `STOP_COMPILATION`). In that case we might unnecessarily leave at most one compilation task in the compilation queue, which we could actually remove. But that doesn't do any harm. The compilation will just fail because of the full code cache with the only draw-back that we've unnecessarily wasted some CPU cycles for the compilation.